enterprise/backend/src/metabase_enterprise/caching/strategies.clj

```diff
@@ -0,0 +1,204 @@
+
+(defmethod fetch-cache-stmt-ee* :nocache [_ _ _]
+  nil)
+
+(defenterprise fetch-cache-stmt
+  "Returns prepared statement to query for db cache backend."
+  :feature :cache-granular-controls
+  [strategy query-hash conn]
+  (fetch-cache-stmt-ee* strategy query-hash conn))
+
+;;; Jobs
+
+(defn- refresh-schedule-configs []
+  (let [now (t/offset-date-time)]
+    (doseq [item (t2/select :model/CacheConfig :strategy :schedule {:where [:or
+                                                                            [:= :next_run_at nil]
+                                                                            [:<= :next_run_at [:now]]]})
+            :let [cron     (-> (cron/cron-schedule (-> item :config :schedule))
+                               cron/finalize)
+                  ;; needed by the tests, or the cron will use it's own current date
+                  _        (.setStartTime ^MutableTrigger cron (t/java-date))
+                  next-run (-> (.getFireTimeAfter ^MutableTrigger cron (t/java-date (t/offset-date-time)))
+                               (t/offset-date-time (t/zone-offset)))]]
+      (t2/update! :model/CacheConfig {:id (:id item)} {:updated_at  now
+                                                       :next_run_at next-run}))))
+
+(defn- refresh-query-configs []
+  (let [now     (t/offset-date-time)
+        configs (t2/select :model/CacheConfig :strategy :query {:where [:or
+                                                                        [:= :next_run_at nil]
+                                                                        [:<= :next_run_at [:now]]]})
+        fields  (m/index-by :id
+                            (t2/select :model/Field :id [:in (map #(-> % :config :field_id) configs)]))
+        tables  (m/index-by :id
+                            (t2/select :model/Table :id [:in (map :table_id (vals fields))]))]
+    (doseq [item configs
+            :let [config   (:config item)
+                  field    (get fields (:field_id config))
+                  table    (get tables (:table_id field))
+                  cron     (-> (cron/cron-schedule (:schedule config))
+                               cron/finalize)
+                  ;; needed by the tests, or the cron will use it's own current date
+                  _        (.setStartTime ^MutableTrigger cron (t/java-date))
+                  next-run (-> (.getFireTimeAfter ^MutableTrigger cron (t/java-date (t/offset-date-time)))
+                               (t/offset-date-time (t/zone-offset)))
+                  query    {:database (:db_id table)
+                            :type     :query
+                            :query    {:source-table (:table_id field)
+                                       :aggregation  [(:aggregation config)
+                                                      [:field (:id field) nil]]}}
+                  result   (-> (qp/process-query query) :data :rows ffirst)]]
+      (when (not= result (:payload item))
```

Comment:
It feels odd for the "next run at" time to fall into the past when the payload hasn't changed.

Reply:
It's a mistake in logic and I should write a test for that.

Reply:
fixed